### PR TITLE
Add currency exchange rates to BlockContext

### LINKED ISSUE #58: Add BaseGasPrice and exchange rates per currency to block context
Subtask of #6 

For comparing legacy and celo transactions, we need access to the base gas prices and exchange rates. They should be added to the block context and only be created once per block.
This might partly be done or required for #53 .

```[tasklist]
### Tasks
- [x] Decide whether to add this directly to the `BlockContext` or add an additional struct like `FeeCurrencyContext`
- [x] Add this to the block context
- [x] Make sure this context is set up at block creation
```

### PR DESCRIPTION
**Pull request by karlb**
I did not add currency specific base fees, as these are implied by the CELO base fee and the exchange rate.

The FeeCurrency is from the unmerged and untested https://github.com/celo-org/optimism/pull/71, but I wanted to include all necessary contracts in one go. We can easily update it later.

The manual initialization of contracts in `core/celo_genesis.go` is a bit fiddly, but IMO still better than pulling in additional tooling.

Since the fee currency support is incomplete, the e2e testing does not actually test anything at the moment. It provides good examples what to run to check the partial support, though.

Closes https://github.com/celo-org/optimism/issues/58

### CODE REVIEW

File: core/celo_evm.go [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+	"github.com/ethereum/go-ethereum/contracts/celo/abigen"
+	contracts_config "github.com/ethereum/go-ethereum/contracts/config"
+	"github.com/ethereum/go-ethereum/core/state"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/core/vm"
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/ethereum/go-ethereum/params"
+)
+
+func addCeloFieldsToBlockContext(blockContext *vm.BlockContext, header *types.Header, config *params.ChainConfig, statedb *state.StateDB) {
```
Comment (palango) [inline core/celo_evm.go]: nit: Maybe call this `setCeloFieldsInBlockContext`

File: core/celo_evm.go [hunk 2]
```diff
@@ -11,6 +15,47 @@ import (
...
+		if err != nil {
+			log.Error("Failed to get medianRate for gas currency!", "err", err, "tokenAddress", tokenAddress)
+			continue
+		}
+		if denominator.Sign() == 0 {
+			log.Error("Bad exchange rate for fee currency", "tokenAddress", tokenAddress, "numerator", numerator, "denominator", denominator)
+			continue
+		}
+		exchangeRates[tokenAddress] = big.NewRat(numerator.Int64(), denominator.Int64())
+		// fmt.Printf("exchange rate for token %v: %v\n", tokenAddress, exchangeRates[tokenAddress])
```
Comment (palango) [inline core/celo_evm.go]: Please either remove this or make it a log call in with a low level.

File: core/celo_genesis.go [hunk 3]
```diff
@@ -49,7 +56,27 @@ func celoGenesisAccounts() map[common.Address]GenesisAccount {
 	if err != nil {
 		panic(err)
 	}
+	sortedOraclesBytecodeLinked := bytes.Replace(contracts.SortedOraclesBytecodeRaw, []byte("__$c0b499b413513d0c67e2a6a17d90846cb3$__"), []byte("000000000000000000000000000000000000ce17"), -1)
```
Comment (palango) [inline core/celo_genesis.go:59]: Will we have to adapt this every time the bytecode changes?

Comment (karlb) [reply core/celo_genesis.go:59]: I failed at looking up that information. I assume it is something stable like `keccak("AddressSortedLinkedListWithMedian")` to get a fixed length placeholder.

**Review by palango — APPROVED** [review]
🚀

**Pull request merged**